File: spec.py
from typing import * # type: ignore
from uuid import uuid4
from pydantic import BaseModel, Field, TypeAdapter

# pyright: reportIncompatibleVariableOverride=false

AnyGameCommand = Literal[
    "startup",
    "context",
    "actions/register",
    "actions/unregister",
    "actions/force",
    "action/result",
    "shutdown/ready",
]

AnyNeuroCommand = Literal[
    "action",
    "actions/reregister_all",
    "shutdown/graceful",
    "shutdown/immediate",
]

class ActionModel(BaseModel):
    name: str
    description: str
    schema: Mapping[str, Any] | None # type: ignore

class Message(BaseModel):
    command: AnyGameCommand | AnyNeuroCommand

class GameMessage(Message):
    command: AnyGameCommand
    game: str

class NeuroMessage(Message):
    command: AnyNeuroCommand

# Game messages
class Startup(GameMessage):
    command: Literal["startup"] = "startup"

class Context(GameMessage):
    class Data(BaseModel):
        message: str
        silent: bool
    
    command: Literal["context"] = "context"
    data: Data

class RegisterActions(GameMessage):
    class Data(BaseModel):
        actions: list[ActionModel]
    
    command: Literal["actions/register"] = "actions/register"
    data: Data

class UnregisterActions(GameMessage):
    class Data(BaseModel):
        action_names: list[str]
    
    command: Literal["actions/unregister"] = "actions/unregister"
    data: Data

class ForceAction(GameMessage):
    class Data(BaseModel):
        state: str | None
        query: str
        ephemeral_context: bool | None
        action_names: list[str]
        main_thread: bool = True # undocumented
    
    command: Literal["actions/force"] = "actions/force"
    data: Data

class ActionResult(GameMessage):
    class Data(BaseModel):
        id: str
        success: bool
        message: str | None
    
    command: Literal["action/result"] = "action/result"
    data: Data

# Neuro messages

class Action(NeuroMessage):
    class Data(BaseModel):
        id: str = Field(default_factory=lambda: uuid4().hex)
        name: str
        data: Any | None
    
    command: Literal["action"] = "action"
    data: Data

# active proposals - see https://github.com/VedalAI/neuro-game-sdk/blob/main/API/PROPOSALS.md

class ReregisterAllActions(NeuroMessage):
    command: Literal["actions/reregister_all"] = "actions/reregister_all"

class GracefulShutdown(NeuroMessage):
    # might be changed; see https://github.com/VedalAI/neuro-game-sdk/issues/44
    class Data(BaseModel):
        wants_shutdown: bool

    command: Literal["shutdown/graceful"] = "shutdown/graceful"
    data: Data

class ImmediateShutdown(NeuroMessage):
    command: Literal["shutdown/immediate"] = "shutdown/immediate"

class ShutdownReady(GameMessage):
    command: Literal["shutdown/ready"] = "shutdown/ready"

AnyGameMessage = Union[Startup, Context, RegisterActions, UnregisterActions, ForceAction, ActionResult, ShutdownReady]
GameMessageAdapter = TypeAdapter(Annotated[AnyGameMessage, Field(discriminator="command")])

AnyNeuroMessage = Union[Action, ReregisterAllActions, GracefulShutdown, ImmediateShutdown]
NeuroMessageAdapter = TypeAdapter(Annotated[AnyNeuroMessage, Field(discriminator="command")])

File: src/lib/api/v1/spec.ts
import { v4 as uuid4 } from "uuid";
import * as z from 'zod';
import { zConst } from "$lib/app/utils";

export const zAction = z.strictObject({
    name: z.string().nonempty(),
    description: z.string().optional(),
    schema: z.record(z.string(), z.any()).nullable(),
});

export type Action = z.infer<typeof zAction>;

export const zGameMessageBase = z.strictObject({
    game: z.string().nonempty(),
    // *technically* an error to include this in dataless messages, but we ball
    // (e.g. Abandoned Pub sends `startup` with `data: {}`)
    data: z.object().nullish(),
});

export const zStartup = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("startup"),
});

export const zContext = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("context"),
    data: z.strictObject({
        message: z.string(),
        silent: z.boolean(),
    }),
});

export const zRegisterActions = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("actions/register"),
    data: z.strictObject({
        actions: z.array(zAction),
    }),
});

export const zUnregisterActions = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("actions/unregister"),
    data: z.strictObject({
        action_names: z.array(z.string()),
    }),
});

export const zForceAction = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("actions/force"),
    data: z.strictObject({
        state: z.string().optional(),
        query: z.string(),
        ephemeral_context: z.boolean().default(false),
        action_names: z.array(z.string()),
        priority: z.enum(["low", "medium", "high", "critical"]).fallback("low"), // TODO: interrupt llm gen (when we start streaming responses)
    }),
});

export const zActionResult = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("action/result"),
    data: z.strictObject({
        id: z.string(),
        success: z.boolean(),
        message: z.string().optional(),
    }),
});

export const zShutdownReady = z.strictObject({
    ...zGameMessageBase.shape,
    command: zConst("shutdown/ready"),
});

export const zGameMessage = z.discriminatedUnion("command", [
    zStartup,
    zContext,
    zRegisterActions,
    zUnregisterActions,
    zForceAction,
    zActionResult,
    zShutdownReady,
]);

export type Startup = z.infer<typeof zStartup>;
export type Context = z.infer<typeof zContext>;
export type RegisterActions = z.infer<typeof zRegisterActions>;
export type UnregisterActions = z.infer<typeof zUnregisterActions>;
export type ForceAction = z.infer<typeof zForceAction>;
export type ActionResult = z.infer<typeof zActionResult>;
export type ShutdownReady = z.infer<typeof zShutdownReady>;

export type GameMessage = z.infer<typeof zGameMessage>;

// Neuro messages

export const zActData = z.strictObject({
    id: z.string().default(uuid4),
    name: z.string().nonempty(),
    data: z.string().nullish(),
});
export const zAct = z.strictObject({
    command: zConst("action"),
    data: zActData,
});

export const zReregisterAll = z.strictObject({
    command: zConst("actions/reregister_all"),
})

export const zNeuroMessage = z.discriminatedUnion("command", [
    zAct,
    zReregisterAll,
])

export type Act = z.infer<typeof zAct>;
export type ActData = z.infer<typeof zActData>;
export type ReregisterAll = z.infer<typeof zReregisterAll>;
export type NeuroMessage = z.infer<typeof zNeuroMessage>;
